File: design_patterns_101/Structural Patterns/Decorator Pattern/decoratorPattern.py
from abc import ABC, abstractmethod

# Component interface
class Component(ABC):
    @abstractmethod
    def operation(self) -> str:
        pass

# Concrete Component
class ConcreteComponent(Component):
    def operation(self) -> str:
        return "ConcreteComponent operation"

# Decorator
class Decorator(Component, ABC):
    def __init__(self, component: Component) -> None:
        self._component = component

    @abstractmethod
    def operation(self) -> str:
        pass

# Concrete Decorators
class ConcreteDecoratorA(Decorator):
    def operation(self) -> str:
        return f"ConcreteDecoratorA({self._component.operation()})"

class ConcreteDecoratorB(Decorator):
    def operation(self) -> str:
        return f"ConcreteDecoratorB({self._component.operation()})"

# Usage
simple = ConcreteComponent()
print("Client: I've got a simple component:")
print(simple.operation())

decorator1 = ConcreteDecoratorA(simple)
decorator2 = ConcreteDecoratorB(decorator1)
print("\nClient: Now I've got a decorated component:")
print(decorator2.operation())

File: design_patterns_101/Structural Patterns/Decorator Pattern/decoratorPattern.js
// Component interface
class Component {
    operation() {
      throw new Error("Abstract method!");
    }
  }
  
  // Concrete Component
  class ConcreteComponent extends Component {
    operation() {
      return "ConcreteComponent operation";
    }
  }
  
  // Decorator
  class Decorator extends Component {
    constructor(component) {
      super();
      this.component = component;
    }
  
    operation() {
      return this.component.operation();
    }
  }
  
  // Concrete Decorators
  class ConcreteDecoratorA extends Decorator {
    operation() {
      return `ConcreteDecoratorA(${this.component.operation()})`;
    }
  }
  
  class ConcreteDecoratorB extends Decorator {
    operation() {
      return `ConcreteDecoratorB(${this.component.operation()})`;
    }
  }
  
  // Usage
  const simple = new ConcreteComponent();
  console.log("Client: I've got a simple component:");
  console.log(simple.operation());
  
  const decorator1 = new ConcreteDecoratorA(simple);
  const decorator2 = new ConcreteDecoratorB(decorator1);
  console.log("\nClient: Now I've got a decorated component:");
  console.log(decorator2.operation());
  